DOC/man/checkproof.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH CHECKPROOF "1" "May 2024" "checkproof 3.1.0" "User Commands"
.SH NAME
checkproof \- manual page for checkproof 3.1.0
.SH SYNOPSIS
.B checkproof
[\fI\,options\/\fR] [\fI\,files\/\fR]
.SH DESCRIPTION
checkproof 3.1.0
.PP
Read an UPCL2 protocol and verify the inferences using one of a
varity of external provers.
.PP
This is a _very_ experimental program. Passing checkproof does
indicate that all inferences in an UPCL2 protocol are correct
(i.e. that the conclusion is logically implied by the premisses) \-
that is, if you believe that the transformation process and the used
prover are correct. However, checkproof will e.g. gladly show that the
empty proof protocol does not contain any buggy steps.
.PP
If a proof protocol fails to pass this test, the proof may still be
correct. Due to e.g. incomplete strategies (this applies in particular
to Otter), build\-in limits (Otter), and bugs in the prover (potentially
all systems, but observed in SPASS 0.55), a prover might fail to
verify a correct step. Moreover, due to the different strategies,
calculi, and in particular different term orderings chosen by the
systems, a single UPCL2 inference may result in a proof problem that
is very hard to verify for other provers. However, if a proof step is
rejected by more than one system, you should probably look at this
step in detail.
.PP
Options
.HP
\fB\-h\fR
.HP
\fB\-\-help\fR
.IP
Print a short description of program usage and options.
.HP
\fB\-\-version\fR
.IP
Print the version number of the program.
.HP
\fB\-v\fR
.HP
\fB\-\-verbose[=\fR<arg>]
.IP
Verbose comments on the progress of the program. The short form or the
long form without the optional argument is equivalent to \fB\-\-verbose\fR=\fI\,1\/\fR.
.HP
\fB\-o\fR <arg>
.HP
\fB\-\-output\-file=\fR<arg>
.IP
Redirect output into the named file.
.HP
\fB\-s\fR
.HP
\fB\-\-silent\fR
.IP
Equivalent to \fB\-\-output\-level\fR=\fI\,0\/\fR.
.HP
\fB\-l\fR <arg>
.HP
\fB\-\-output\-level=\fR<arg>
.IP
Select an output level, greater values imply more verbose output. At the
moment, level 0 only prints the result, level 1 prints inference steps as
they are verified, level 2 prints prover commands issued, and level 3
prints all prover output (which may be very little)
.HP
\fB\-p\fR <arg>
.HP
\fB\-\-prover\-type=\fR<arg>
.IP
Set the type of the prover to use for proof verification. Determines
problem syntax, options, and check for success. Supported options at are
\&'E' (the default),'Otter' 'SPASS', and 'scheme\-setheo' (not yet
implemented). SPASS support is only tested with SPASS 0.55 and may fail
if the problem contains identifiers reserved by SPASS. There have been
some supple syntax changes, so more recent SPASS versions will probably
fail as well.
.HP
\fB\-x\fR <arg>
.HP
\fB\-\-executable=\fR<arg>
.IP
Give the name under which the prover can be called. If no executable is
given, checkproof will guess a name based on the type of the prover. This
guess may be way off!
.HP
\fB\-t\fR <arg>
.HP
\fB\-\-prover\-cpu\-limit=\fR<arg>
.IP
Limit the CPU time prover may spend on a single step. Default is 10
seconds.
.SH "REPORTING BUGS"
.PP
Report bugs to <schulz@eprover.org>. Please include the following, if
possible:
.PP
* The version of the package as reported by \fBeprover \-\-version\fR.
.PP
* The operating system and version.
.PP
* The exact command line that leads to the unexpected behaviour.
.PP
* A description of what you expected and what actually happened.
.PP
* If possible all input files necessary to reproduce the bug.
.SH COPYRIGHT
Copyright 1998\-2024 by Stephan Schulz, schulz@eprover.org,
and the E contributors (see DOC/CONTRIBUTORS).
.PP
This program is a part of the distribution of the equational theorem
prover E. You can find the latest version of the E distribution
as well as additional information at
http://www.eprover.org
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program (it should be contained in the top level
directory of the distribution in the file COPYING); if not, write to
the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
Boston, MA  02111\-1307 USA
.PP
The original copyright holder can be contacted via email or as
.PP
Stephan Schulz
DHBW Stuttgart
Fakultaet Technik
Informatik
Lerchenstrasse 1
70174 Stuttgart
Germany


DOC/NEWS
This history describes the major differences between versions of the E
equational theorem prover.

Version 3.1.0 -> 3.2.X Puttabong Moondrop

- Added support for $distinct
- Disabled -xAuto and cut out most of the relics of the old auto-mode
  stuff
- Added option to select one of the built-in strategies
- Took out old clausifier
- Updated several options
- Debugged some higher-order problems

Version 3.0.03 -> 3.1.0

- Added disequality decomposition inference
- Fixed various bugs (e.g. in local rewriting)
- Added support for additional statistics
- Enabled printout of all built-in search strategies


Version 3.0 ->3.0.03 Shangri-La

- Fixed bug in activating clause selection heuristics
- Fixed several bugs in the higher-order module, mostly related to the
  difference between shared and unshared terms. This will need further
  work to catch all corner cases.
- Fixed several minor memory leaks and simplified management of parsed
  strings with the clb_permastrings module.

Version 3.0 (pre-release) ->3.0 Shangri-La

- Fixed various bugs in the HO preprocessor
- Fixed various bugs in the HO reasoner
- Improved the multicore scheduler

Version 2.6 ->3.0 Shangri-La (pe-release)

- New option --fw-subsumption-aggressive.
- Fixed bug in negative unit subsumption.
- Added new DAG-based evaluation functions
- Fixed status for incomplete literal weight functions
- Added support for UTF-8 in Strings and single-quoted symbols.
- Added multicore scheduler and schedule generation
- Replaced auto-mode code by new symbolic auto-mode
- Various minor improvements
- Added support for full (monomorphic) higher-order reasoning (PV)


Version 2.5 -> 2.6 Floral Guranse

- Much improved algorithm for finding maximal literals in long clauses
- Much improved algorithm for identifying dual (conflicting) literals
  in long clauses
- Added max literal limit for full equational tautology check (we
  should now really replace that with a congruence closure based
  approach!)
- Added DAGWeight() evaluation function (symbol-counts shared terms
  only once, with different options for sharing regions (terms,
  literals, clauses)
- Improved support for THF: E can now parse most of TH0 syntax
- (notable exceptions are symbols !!, ?? and @+), including lambdas.
  Note that on the reasoning side, E still does not support any
  advanced lambda reasoning, as lambdas are simply lifted and after
  parsing E is not aware of them.
- Improved support for TFX: E now supports $let and $ite expressions.



Version 2.4 -> 2.5 Avongrove

- Clausal problems are now classified before equational definitions
  expansion, which moves equational definition expansion under control
  auf the automatic modes)
- Strong rewrite (optionally instantiating unbound variables of the
  right hand side of equations considered for rewriting) has neem
  fixed and updeted to the sorted case.
- An improvement in strategy scheduling will now aovid repetition of
  the same strategy in cases where the data does not change from one
  level to the next, allowing for meaningful deeper schedules.
- New automatic modes
- Added variants to symbol precedence schemes (prefer/defer
  skolems/defined symbols/axiom-symbols/conjecture symbols).


Version 2.3 -> 2.4 Sandakphu

- Added random clause "evaluation"
- Added Diversity-based clause evaluation
- Added (simplified) transfinite orderings for literal comparision
- New automatic modes
- Support for CASC-27 LTB (with various alternative formalisations per
  problem)
- Slightly finer classificarion of problems (for improved UEQ
  performance)

Version 2.3pre -> 2.3 Gielle

- Made clausification more robust for extreme examples

Version 2.2 -> 2.3pre Gielle

- Merged support for lambda-free higher-order logic.
- Improved fingerprint indexing by excluding one more subcase for
  finding instances

Version 2.1 -> 2.2 Thurbo Moonlight

- Prover now always builds internal proof object (--proof-object only
  controls output)
- As a result, E can now detect ContradictoryAxioms if the proof state
  is unsatisfiable even without the conjecture.
- Moved from eager to lazy orphan deletion. If you don't know what
  this means: It's much cooler and simpler.
- Fixed and simplified processing of unprocessed clauses
- Refactored PicoSAT integration - PicoSAT is now linked into E
  proper, and used via its API. We are back to a single binary
  executable with no dependencies.
- Prover now enforces that fof, tff and tcf formulas are closed (no
  free variables allowed)
- Fixed condensation (somewhat).
- New automatic modes (but with very few current test results)


Version 2.0 -> 2.1 Maharani Hills

- Added type output for CNFed TFFs
- Added tcf sublanguage for clause-form TFFs
- Combined consecutive applications of the same quantifier into a
  single quantifier
- Improved stability and performance on TFF problems
- Changed internal workflow of unprocessed clauses so that they can be
  batch-evaluated (e.g. by a Deep Neural Network).
- Added PicoSAT and the ability to use PicoSAT as a SAT terminator
- Added support for flexible-arity derivation operations
- Removed obsolete eproof-scripts. Use --proof-object to generate a
  checkable proof.

Version 1.9.1 -> 2.0 Turzum

- Support for TFF (typed first-order form), thanks to Simon Cruanes
- Improved clausification (faster, with graceful degradation if
  miniscoping becomes too expensive)
- Improved auto-modes
- Improved Set-of-Support simulation
- Parsing and pruning preprocessing FOF now, even for CNF input
- Input clauses/formulas keep their original name (if any)
- Added options  --generated-limit and --tb-insert-limit to more
  closely limit effort by the prover in a deterministic way.

Version 1.9 -> 1.9.1 Sungma

- Improved proof object output
- Added optional recording of selected given clauses in the proof
  object
- New auto modes (including one bred heuristic)
- Fixed bug in formula simplification involving P=><=>$false

PROVER/checkproof.c
/*-----------------------------------------------------------------------

File  : checkproof.c

Author: Stephan Schulz

Contents

  Read a PCL protocol and try to verify it using a selected prover.

  Copyright 1998, 1999 by the author.
  This code is released under the GNU General Public Licence and
  the GNU Lesser General Public License.
  See the file COPYING in the main E directory for details..
  Run "eprover -h" for contact information.

Changes

<1> Fri Apr  7 16:14:02 MET DST 2000
    New

-----------------------------------------------------------------------*/

#include <stdio.h>
#include <cio_commandline.h>
#include <cio_output.h>
#include <cio_tempfile.h>
#include <cio_signals.h>
#include <pcl_proofcheck.h>
#include <e_version.h>


/*---------------------------------------------------------------------*/
/*                  Data types                                         */
/*---------------------------------------------------------------------*/

#define NAME    "checkproof"

typedef enum
{
   OPT_NOOPT=0,
   OPT_HELP,
   OPT_VERSION,
   OPT_VERBOSE,
   OPT_OUTPUT,
   OPT_SILENT,
   OPT_OUTPUTLEVEL,
   OPT_PROVERTYPE,
   OPT_EXECUTABLE,
   OPT_TIME_LIMIT
}OptionCodes;



/*---------------------------------------------------------------------*/
/*                        Global Variables                             */
/*---------------------------------------------------------------------*/

OptCell opts[] =
{
   {OPT_HELP,
    'h', "help",
    NoArg, NULL,
    "Print a short description of program usage and options."},

   {OPT_VERSION,
    '\0', "version",
    NoArg, NULL,
    "Print the version number of the program."},

   {OPT_VERBOSE,
    'v', "verbose",
    OptArg, "1",
    "Verbose comments on the progress of the program."},

   {OPT_OUTPUT,
    'o', "output-file",
    ReqArg, NULL,
   "Redirect output into the named file."},

   {OPT_SILENT,
    's', "silent",
    NoArg, NULL,
    "Equivalent to --output-level=0."},

   {OPT_OUTPUTLEVEL,
    'l', "output-level",
    ReqArg, NULL,
    "Select an output level, greater values imply more verbose"
    " output. At the moment, level 0 only prints the result, level 1"
    " prints inference steps as they are verified, level 2 prints"
    " prover commands issued, and level 3 prints all prover output"
    " (which may be very little)"},

   {OPT_PROVERTYPE,
    'p', "prover-type",
    ReqArg, NULL,
    "Set the type of the prover to use for proof"
    " verification. Determines problem syntax, options, and check for"
    " success. Supported options at are  'E' (the"
    " default),'Otter' 'SPASS', and 'scheme-setheo' (not yet"
    " implemented). SPASS support is only tested with SPASS"
    " 0.55 and may fail if the problem contains identifiers reserved "
    "by SPASS. There have been some supple syntax changes, so more"
    " recent SPASS versions will probably fail as well."},

   {OPT_EXECUTABLE,
    'x', "executable",
    ReqArg, NULL,
    "Give the name under which the prover can be called. If no"
    " executable is given, checkproof will guess a name based on the"
    " type of the prover. This guess may be way off!"},

   {OPT_TIME_LIMIT,
    't', "prover-cpu-limit",
    ReqArg, NULL,
    "Limit the CPU time prover may spend on a single step. Default is"
    " 10 seconds."},

   {OPT_NOOPT,
    '\0', NULL,
    NoArg, NULL,
    NULL}
};

char       *outname    = NULL;
long       time_limit  = 10;
char       *executable = NULL;
ProverType prover      = EProver;
bool       app_encode  = false;


/*---------------------------------------------------------------------*/
/*                      Forward Declarations                           */
/*---------------------------------------------------------------------*/

CLState_p process_options(int argc, char* argv[]);
void print_help(FILE* out);

/*---------------------------------------------------------------------*/
/*                         Internal Functions                          */
/*---------------------------------------------------------------------*/


int main(int argc, char* argv[])
{
   CLState_p   state;
   Scanner_p   in;
   PCLProt_p   prot;
   long        steps,res,unchecked;
   int         i;

   assert(argv[0]);

   InitIO(NAME);
   atexit(TempFileCleanup);

   ESignalSetup(SIGTERM);
   ESignalSetup(SIGINT);

   OutputFormat = TPTPFormat;
   /* We need consistent name->var mappings here because we
      potentially read the compressed input format. */
   ClausesHaveLocalVariables = false;

   state = process_options(argc, argv);

   OpenGlobalOut(outname);
   prot = PCLProtAlloc();

   if(state->argc ==  0)
   {
      CLStateInsertArg(state, "-");
   }
   steps = 0;
   for(i=0; state->argv[i]; i++)
   {
      in = CreateScanner(StreamTypeFile, state->argv[i], true, NULL, true);
      ScannerSetFormat(in, TPTPFormat);
      steps+=PCLProtParse(in, prot);
      CheckInpTok(in, NoToken);
      DestroyScanner(in);
   }
   VERBOUT2("PCL input read\n");

   res = PCLProtCheck(prot,prover,executable,time_limit, &unchecked);

   fprintf(GlobalOut,
         "# Successfully checked %ld of %ld steps (%ld unchecked): ",
         res, steps, unchecked);

   if(res==steps)
   {
      fprintf(GlobalOut, " Proof verified!\n");
   }
   else if((res+unchecked) == steps)
   {
      fprintf(GlobalOut, " Proof partially verified!\n");
   }
   else
   {
      fprintf(GlobalOut,
         " Failed to verify proof!\n");
   }

   PCLProtFree(prot);

   CLStateFree(state);

   fflush(GlobalOut);
   OutClose(GlobalOut);
   ExitIO();

#ifdef CLB_MEMORY_DEBUG
   MemFlushFreeList();
   MemDebugPrintStats(stdout);
#endif

   return 0;
}


/*-----------------------------------------------------------------------
//
// Function: process_options()
//
//   Read and process the command line option, return (the pointer to)
//   a CLState object containing the remaining arguments.
//
// Global Variables:
//
// Side Effects    : Sets variables, may terminate with program
//                   description if option -h or --help was present
//
/----------------------------------------------------------------------*/

CLState_p process_options(int argc, char* argv[])
{
   Opt_p handle;
   CLState_p state;
   char*  arg;

   state = CLStateAlloc(argc,argv);

   while((handle = CLStateGetOpt(state, &arg, opts)))
   {
      switch(handle->option_code)
      {
      case OPT_VERBOSE:
       Verbose = CLStateGetIntArg(handle, arg);
       break;
      case OPT_HELP:
       print_help(stdout);
       exit(NO_ERROR);
      case OPT_VERSION:
       printf(NAME " " VERSION "\n");
       exit(NO_ERROR);
      case OPT_OUTPUT:
       outname = arg;
       break;
      case OPT_SILENT:
       OutputLevel = 0;
       break;
      case OPT_OUTPUTLEVEL:
       OutputLevel = CLStateGetIntArg(handle, arg);
       break;
      case OPT_PROVERTYPE:
       if(strcmp(arg, "E")==0)
       {
          prover = EProver;
       }
       else if(strcmp(arg, "Otter")==0)
       {
          prover = Otter;
          OutputFormat = LOPFormat;
          EqnUseInfix = false;
       }
       else if(strcmp(arg, "SPASS")==0)
       {
          prover = Spass;
          OutputFormat = LOPFormat;
          EqnUseInfix = false;
       }
       else if(strcmp(arg, "scheme-setheo")==0)
       {
          prover = Setheo;
       }
       else
       {
          Error("Option -p (--prover-type) requires E, "
           "Otter, SPASS or scheme-setheo as an argument",
           USAGE_ERROR);
       }
       break;
      case OPT_EXECUTABLE:
       executable = arg;
       break;
      case OPT_TIME_LIMIT:
       time_limit = CLStateGetIntArg(handle, arg);
       break;
      default:
    assert(false);
    break;
      }
   }
   return state;
}

void print_help(FILE* out)
{
   fprintf(out, "\n\
\n"
NAME " " VERSION "\n\
\n\
Usage: " NAME " [options] [files]\n\
\n\
Read an UPCL2 protocol and verify the inferences using one of a\n\
varity of external provers.\n\
\n\
This is a _very_ experimental program. Passing checkproof does\n\
indicate that all inferences in an UPCL2 protocol are correct\n\
(i.e. that the conclusion is logically implied by the premisses) -\n\
that is, if you believe that the transformation process and the used\n\
prover are correct. However, checkproof will e.g. gladly show that the\n\
empty proof protocol does not contain any buggy steps.\n\
\n\
If a proof protocol fails to pass this test, the proof may still be\n\
correct. Due to e.g. incomplete strategies (this applies in particular\n\
to Otter), build-in limits (Otter), and bugs in the prover (potentially\n\
all systems, but observed in SPASS 0.55), a prover might fail to\n\
verify a correct step. Moreover, due to the different strategies,\n\
calculi, and in particular different term orderings chosen by the\n\
systems, a single UPCL2 inference may result in a proof problem that\n\
is very hard to verify for other provers. However, if a proof step is\n\
rejected by more than one system, you should probably look at this\n\
step in detail.\n\
\n");
   PrintOptions(stdout, opts, "Options\n\n");
   fprintf(out, "\n\n" E_FOOTER);
}


/*---------------------------------------------------------------------*/
/*                        End of File                                  */
/*---------------------------------------------------------------------*/
